### PR TITLE
[JENKINS-71692] Attempt to stop step executions if the CPS VM thread dies with a fatal error

### PR DESCRIPTION
**Pull request by dwnusbaum**
See [JENKINS-71692](https://issues.jenkins.io/browse/JENKINS-71692).

If there is a fatal error in the CPS VM thread, currently running steps should be given a chance to stop before we kill the build, otherwise async steps will just keep running in the background. Things can get very messy since the execution and run associated with the `StepExecution` will already be dead, and some steps like `node` and `stage` will just run forever if we do not stop them.

### Testing done

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [ ] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

### CODE REVIEW

File: plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java [hunk 1]
```diff
@@ -53,8 +56,34 @@ public void run() {
...
+                            listener.getLogger().println("Terminating " + node.getDisplayFunctionName() + " (id: " + node.getId() + ")");
+                        }
+                    } catch (Exception e) {
+                        t.addSuppressed(e);
+                    }
+                }
+            }
+        } catch (Throwable t2) {
+            t.addSuppressed(t2);
+        }
```
Comment (dwnusbaum) [inline plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java:84]: I think if we switch the inner catch to `Throwable` this should be unreachable and could be deleted.

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java [hunk 2]
```diff
@@ -664,4 +667,47 @@ public boolean isAllowed(String groovyResourceUrl) {
...
+                "          semaphore('croak')\n" +
+                "          break label\n" + // CallEnv.getBreakAddress throws an IllegalStateException when this runs.
+                "        }\n" +
+                "      }\n" +
+                "    }\n" +
+                "  },\n" +
+                "  'keepsRunning': {\n" +
+                "    semaphore('other')\n" +
+                "    echo 'kept running!'\n" +
+                "  }\n" +
```
Comment (dwnusbaum) [inline plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java:689]: Not necessary to reproduce the main problem with `node`, but this shows how steps could still "run" even after `CpsFlowExecution.croak`.

File: plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java (hunk 1 above)
Comment (dwnusbaum) [reply plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java]: Simplified in [e76ebb6](https://github.com/jenkinsci/workflow-cps-plugin/pull/780/commits/e76ebb65eb3178d6a21599ce23d72ac616dec9d8).

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java [hunk 3]
```diff
@@ -664,4 +670,117 @@ public boolean isAllowed(String groovyResourceUrl) {
...
+            // It's difficult to add meaningful non-flaky test assertions here.
+            // Even before the associated fix, 'kept running' wasn't printed, because although CpsThreadGroup.run
+            // did execute again, it hit an IllegalStateException in SandboxContinuable, which resulted in a second
+            // call to CpsFlowExecution.croak.
+            assertTrue(((CpsFlowExecution) b.getExecution()).programPromise.get(1, TimeUnit.SECONDS).runner.isShutdown());
+        });
+    }
+
+    @Issue("JENKINS-71692")
+    @Test public void stepsAreStoppedWhenBodyExecutionCallbackThrows() throws Throwable {
```
Comment (dwnusbaum) [inline plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java:717]: Mimics the case described in [this comment](https://issues.jenkins.io/browse/JENKINS-70267?focusedId=434633&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-434633), although the callback was also fixed upstream, see https://github.com/jenkinsci/gitlab-plugin/pull/1410.

File: plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java [hunk 4]
```diff
@@ -53,8 +56,30 @@ public void run() {
      * That makes it worth reporting.
      */
     private void reportProblem(Throwable t) {
+        if (isShutdown()) {
```
Comment (dwnusbaum) [inline plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java:59]: Or maybe better to check this instead:
```suggestion
        if (cpsThreadGroup.getExecution().isComplete()) {
```

Comment (car-roll) [reply plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java:59]: Not sure about the micro specifics between the two, but using `getExecution` at least looks more readable to me.

**Review by car-roll — APPROVED** [review]

File: plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java (hunk 4 above)
Comment (dwnusbaum) [reply plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java]: See [b9080e8](https://github.com/jenkinsci/workflow-cps-plugin/pull/780/commits/b9080e88143ad3eb6cdb6e57ad5fdd91abab9ce4).

File: plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java [hunk 5]
```diff
@@ -53,8 +56,35 @@ public void run() {
...
+                    if (listener != null && node != null) {
+                        listener.getLogger().println("Terminating " + node.getDisplayFunctionName() + " (id: " + node.getId() + ")");
+                    }
+                } catch (Throwable e) {
+                    t.addSuppressed(e);
+                }
+            }
+        }
         cpsThreadGroup.getExecution().croak(t);
+        // cpsThreadGroup.run() must not execute again. We shut it down after stopping steps and completing the build
```
Comment (dwnusbaum) [inline plugin/src/main/java/org/jenkinsci/plugins/workflow/cps/CpsVmExecutorService.java:84]:
```suggestion
        // cpsThreadGroup.run() must not execute again. We shut this executor service down after stopping steps and completing the build
```

**Review by car-roll — APPROVED** [review]

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java (hunk 3 above)
Comment (dwnusbaum) [reply plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java:717]: Yes I also tested changes along those lines, but to start with I just wanted to show that the new fix is somewhat generic and covers various kinds of fatal VM thread errors. I will add the special case now though.

Comment (dwnusbaum) [reply plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java:717]: Updated to match `TailCall` behavior in [a27f32f](https://github.com/jenkinsci/workflow-cps-plugin/pull/780/commits/a27f32fd7a35f3e847e10a429cc7c15c723df26f).

File: plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java [hunk 6]
```diff
@@ -664,4 +670,117 @@ public boolean isAllowed(String groovyResourceUrl) {
...
+        sessions.then(r -> {
+            WorkflowJob p = r.createProject(WorkflowJob.class, "p");
+            p.setDefinition(new CpsFlowDefinition(
+                "parallel(\n" +
+                "  'willCroak': {\n" +
+                "    node {\n" +
+                "      label: {\n" + // Defines a label on a BlockStatement, which is erroneously ignored by CpsTransformer.
+                "        while (true) {\n" +
+                "          semaphore('croak')\n" +
+                "          break label\n" + // CallEnv.getBreakAddress throws an IllegalStateException when this runs.
```
[the remark this answers is not in the document]
Comment (dwnusbaum) [reply plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java:684]: Yeah, let me see if we can use `CpsFlowExecution.runInCpsVmThread` to exercise the same general behavior in a way that we will never fix.

[the remark this answers is not in the document]
Comment (dwnusbaum) [reply plugin/src/test/java/org/jenkinsci/plugins/workflow/cps/CpsFlowExecutionTest.java]: Test updated to use `runInCpsVmThread` in [0354b48](https://github.com/jenkinsci/workflow-cps-plugin/pull/780/commits/0354b48597a469e20fd027392a10cdaec0c721a2), and I manually checked that without the fix the updated test triggers the same behavior as with the label bug.

**Review by jglick — APPROVED** [review]

**Pull request merged**